sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/ca/DefaultCAClient.java

```diff
@@ -0,0 +1,159 @@
+        data = new JSONObject(responseContent);
+
+        if (!data.getBoolean("success")) {
+            throw new CAException(getErrorString(data));
+        }
+
+        String certificate = data.getJSONObject("result").getString("certificate");
+
+        return (X509Certificate) certificateFactory
+                .generateCertificate(
+                        new ByteArrayInputStream(certificate.getBytes(StandardCharsets.UTF_8)));
+    }
+
+    @Override
+    public Collection<X509Certificate> chainWithRootCert(
+            X509Certificate certificate) throws Exception {
+        JSONObject data = new JSONObject();
+        data.put("certificate", PEMHelper.toPEM(certificate));
+
+        Request request = Request.Post(URLHelper.addPathUnchecked(baseURL, "bundle").toString())
+                .bodyString(data.toString(), ContentType.APPLICATION_JSON);
+        Response response = httpExecutor.execute(request);
+        HttpResponse httpResponse = response.returnResponse();
+
+        handleResponseStatusLine(httpResponse.getStatusLine(), 200);
+
+        String responseContent = new ContentResponseHandler().handleEntity(httpResponse.getEntity()).asString();
+        data = new JSONObject(responseContent);
+
+        if (!data.getBoolean("success")) {
+            throw new CAException(getErrorString(data));
+        }
+
+        String bundle = data.getJSONObject("result").getString("bundle");
+        ArrayList<X509Certificate> certificates = new ArrayList<>();
+
+        if (bundle.length() > 0) {
+            certificates.addAll(
+                    certificateFactory.generateCertificates(
+                            new ByteArrayInputStream(bundle.getBytes(StandardCharsets.UTF_8)))
+                        .stream()
+                        .map(cert -> (X509Certificate) cert)
+                        .collect(Collectors.toList())
+            );
+            // Bundle response includes also submitted certificate which we don't need
+            // so remove it.
+            certificates.remove(0);
+        }
+
+        // Response should come with Root CA certificate which isn't included in 'bundle'
+        String rootCACert = data.getJSONObject("result").getString("root");
+        if (rootCACert.length() == 0) {
+            throw new CertificateException("Failed to retrieve Root CA certificate");
+        }
+
+        certificates.add((X509Certificate) certificateFactory.generateCertificate(
+                new ByteArrayInputStream(rootCACert.getBytes(StandardCharsets.UTF_8)))
+        );
+
+        return certificates;
+    }
+
+    /**
+     * Handle common responses from different API endpoints of DC/OS CA service.
+     * @param statusLine
+     * @param okCode
+     * @throws CAException
+     */
+    protected void handleResponseStatusLine(StatusLine statusLine, int okCode) throws CAException {
+        if (statusLine.getStatusCode() == okCode) {
```

Comment:
Would it make sense to just throw if `statusLine.getStatusCode() != okCode`? The presence of okCode implies that we know exactly what we expect here.